api/src/main/java/org/apache/iceberg/expressions/Evaluator.java

```diff
@@ -134,13 +135,13 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
     }
 
     @Override
-    public <T> Boolean in(BoundReference<T> ref, Literal<T> lit) {
-      throw new UnsupportedOperationException("In is not supported yet");
+    public <T> Boolean in(BoundReference<T> ref, Set<Literal<T>> lits) {
+      return lits.contains(Literals.from(ref.get(struct)));
```

Comment:
I think it is a problem that this calls `Literals.from` for every comparison. That inspects the value and builds a literal for it in a series of `if` statements. It should be avoided in a tight loop.

Reply:
I think a better approach would be to add a new class, `LiteralSet<T>` that is passed to this method instead of `Set<Literal<T>>`. That would implement `Set<T>` so that `set.contains(ref.get(struct))` works directly. That takes advantage of the fact that we know all of the literals and the value from the struct are already the same type.

Reply:
Thanks for the comment! 
Taking a quick look at the literal's `comparator`s and seems all of them are consistent with `equals` of the underlying values. 
So it seems to be safe to just put the values into the `Set` instead of putting literals.
I will make a change to avoid the unnecessary literal cast.

Reply:
Be careful with strings. You'll probably need to use `Set<CharSequenceWrapper>` so that all `CharSequence` instances are compared correctly.

Reply:
@rdblue  There is an issue with `CharSequence`. For example, `String("abc")` and `UTF8("abc")` are not equal but the comparator returns 0. This causes the `IN` eval return `false`.
To fix it, I created a wrapper class and overrides the `equals` to use the provided comparator. It worked in our cases although it is not perfect as `Set<WrapperA> contains B` but `Set<B> may not contain WrapperA`.

Reply:
Oops, I missed your comment. 
Yep, you are right. I found this issue and came up with a similar solution.
Thanks.